doc/grpc_client/design_doc.md

```diff
@@ -0,0 +1,509 @@
+| 0.1 | 04/1/22 | Vaibhav Dahiya  | Initial version    |
+| 0.2 | 02/1/22 | Vaibhav Dahiya  | Make chnages to be shared to Core Team    |
+
+## Scope
+gRPC client design doc which would communicate with the SoC in DualToR Active-active setup/ and Nic-simulator for testing infrastructure in SONiC MGMT.
+### Overview
+
+This document summarizes the approach taken to accommodate gRPC client
+for DualToR active active scenario. The gRPC client daemon's
+main purpose is to provide a way for linkmgr to exchange RPC's with SoC
+and do this within SONiC PMON docker
+
+
+## Requirements
+
+- provide a service/daemon in SONiC to run in DualToR mode, which can interact with Platform API as well interact with state machine(aka Linkmgr) and orchagent to provide capability for it to get/set Link State/Forwarding State etc. from SoC(gRPC server listening to the client)
+- the service gRPC daemon should be able exchange RPC's with the gRPC server running on the SoC over a secure channel
+- provide a schema for this daemon to publish to State DB on Host which would monitor the aspects of gRPC state for all SoC's running as server.
+- provide an interface/method for gRPC daemon to exchange RPC's with the gRPC server running on the SoC using a loopback IP as source IP.
+- provide an interface for SoC to notify this gRPC client about going to maintainence/shutdown via an asynchronous method.
+- gRPC client communication with Nic-simulator(which will be run in SONiC-Mgmt Testbeds) should also be provided to exchange RPC's.
+- provide a way to monitor gRPC client's and channel health for corrective/monitoring action to be implemented within SONiC ecosystem
+
+
+## whygRPC
+
+## why gRPC for communication between ToR AND the SoC
+
+Notes provide a helpful link for learning gRPC and main page
+
+- Lightweight messages. Depending on the type of call, gRPC-specific messages can be up to 30-50 percent smaller in size than JSON messages.
+- High performance. By different evaluations, gRPC is 5, 7, and even 8 times faster than REST+JSON communication.
+- Built-in code generation. gRPC has automated code generation in different programming languages including Java, C++, Python, Go, Dart, Objective-C, Ruby, and more.
+- More connection options. While REST focuses on request-response architecture, gRPC provides support for data streaming with event-driven architectures: server-side streaming, client-side streaming, and bidirectional streaming.
+- Healthy developer EcoSystem and gRPC is open source, hence it helps in getting acquainted with Libraries/API and Troubleshooting bugs/issues becomes easier
+
+More Resources for learning gRPC and advantages Credits
+
+[grpc github repo](https://github.com/grpc/grpc)
+
+
+## Hardware Overview and overall Archtecture
+
+### Hardware Overview
+
+![Hardware Overview](images/gRPC_overall.png)
+
+### HOST architecture
+
+HOST and FPGA functionality is explained in this diagram
+
+![Hardware Overview](images/gRPC_host.png)

```

Comment:
What is the blue line X for within FPGA indicating?

Reply:
explained offline